tests/test_assignment.py

```diff
@@ -76,3 +88,28 @@ def _prepare_submission(self):
 
         self.assertEqual(total_score, 0)
         self.assertEqual(max_score, 1)
+
+    def test_hard_fail(self):
```

Comment:
I think you need some more test cases here.
In Python, they call them subtests:
https://docs.python.org/3/library/unittest.html#distinguishing-test-iterations-using-subtests

I would just powerset it (all 4 possibilities of a test hard failing).